### PR TITLE
Add Input::request()

### PR DESCRIPTION
**Pull request by Nyholm**
Building on the ideas of #299

### CODE REVIEW

**Comment by Nyholm** [comment]
One could also remove the `requestBody()` but that requires changes in the serializers. I leave that to a second PR.

File: src/CodeGenerator/src/Generator/InputGenerator.php [hunk 1]
```diff
@@ -375,8 +377,45 @@ private function inputClassRequestGetters(StructureShape $inputShape, ClassType
...
+// Return the Request
+return new Request($method, \$uriString, \$query, \$headers, StreamFactory::create($bodyCall));
+PHP
+);
+    }
+
+    /**
+     * Convert variable to a string.
+     */
+    private function stringify(string $variable, Member $member): string
```
Comment (Nyholm) [inline src/CodeGenerator/src/Generator/InputGenerator.php]: I found bugs with our current headers.. They were not properly formatted.

File: src/CodeGenerator/src/Generator/InputGenerator.php [hunk 2]
```diff
@@ -374,9 +376,46 @@ private function inputClassRequestGetters(StructureShape $inputShape, ClassType
...
+    {
+        $type = $member->getShape()->getType();
+        if ('timestamp' === $type) {
+            if ('Expires' === $member->getLocationName()) {
+                return $variable . '->format(\DateTimeInterface::RFC1123)';
+            }
+
+            return $variable . '->format(\DateTimeInterface::ATOM)';
```
Comment (Nyholm) [inline src/CodeGenerator/src/Generator/InputGenerator.php]: Im not 100% about this format. I found some examples but not official

File: src/CodeGenerator/src/Generator/InputGenerator.php [hunk 3]
```diff
@@ -374,9 +376,46 @@ private function inputClassRequestGetters(StructureShape $inputShape, ClassType
...
+        }
+
+        if ('boolean' === $type) {
+            return $variable . ' ? "true" : "false"';
+        }
 
-        $class->addMethod('requestUri')->setComment('@internal')->setReturnType('string')->setBody($requestUri);
+        return $variable;
```
Comment (jderusse) [inline src/CodeGenerator/src/Generator/InputGenerator.php]: I'd add a `if ('string' ...) ` else `throw new InvalidArgumentException('not yet implemented');`

File: src/CodeGenerator/src/Generator/InputGenerator.php [hunk 4]
```diff
@@ -374,9 +376,46 @@ private function inputClassRequestGetters(StructureShape $inputShape, ClassType
...
+    }
+
+    /**
+     * Convert variable to a string.
+     */
+    private function stringifyHeaderValue(string $variable, Member $member): string
+    {
+        $type = $member->getShape()->getType();
+        if ('timestamp' === $type) {
+            if (in_array($member->getLocationName(), ['Expires', 'If-Modified-Since', 'If-Unmodified-Since'])) {
```
[the remark this answers is not in the document]
Comment (Nyholm) [reply src/CodeGenerator/src/Generator/InputGenerator.php]: Thank you. I searched but I didnt find these.

File: src/CodeGenerator/src/Generator/InputGenerator.php (hunk 3 above)
Comment (Nyholm) [reply src/CodeGenerator/src/Generator/InputGenerator.php]: Added!

File: src/Service/S3/src/Input/PutObjectRequest.php [hunk 5]
```diff
@@ -541,7 +536,7 @@ public function requestHeaders(): array
             $headers['Content-Language'] = $this->ContentLanguage;
         }
         if (null !== $this->ContentLength) {
-            $headers['Content-Length'] = $this->ContentLength;
+            $headers['Content-Length'] = (string) $this->ContentLength;
```
Comment (Nyholm) [inline src/Service/S3/src/Input/PutObjectRequest.php]: I make sure headers always are strings and not integers. This is according to RFC 7230

Comment (jderusse) [reply src/Service/S3/src/Input/PutObjectRequest.php]: shouldn't it be done in `Request` object?

**Comment by Nyholm** [comment]
Thank you

File: src/Service/S3/src/Input/PutObjectRequest.php (hunk 5 above)
Comment (Nyholm) [reply src/Service/S3/src/Input/PutObjectRequest.php]: Oops. I merged too fast.. 

Yeah.. maybe.  But the $header annotation on the request object says there should be strings already.

**Pull request merged**